p2p/rlpx.go

```diff
@@ -151,97 +156,236 @@ loop:
 	return status, nil
 }
 
-// ReadAndServe reads messages from peers.
-func (c *Conn) ReadAndServe() *Error {
+// request stores the request ID and the block's hash.
+type request struct {
+	requestID uint64
+	hash      common.Hash
+}
+
+// ReadAndServe reads messages from peers and writes it to a database.
+func (c *Conn) ReadAndServe(db database.Database, count *MessageCount) error {
+	// requests is used to store the request ID and the block hash. This is used
+	// when fetching block bodies because the eth protocol block bodies do not
+	// contain information about the block hash.
+	requests := list.New()
+	var requestNum uint64 = 0
+
+	// dbCh is used to limit the number of database goroutines running at one
+	// time with a buffered channel. Without this, a large influx of messages can
+	// bog down the system and leak memory.
+	var dbCh chan struct{}
+	if db != nil {
+		dbCh = make(chan struct{}, db.MaxConcurrentWrites())
+	}
+
 	for {
 		start := time.Now()
+
 		for time.Since(start) < timeout {
 			if err := c.SetReadDeadline(time.Now().Add(10 * time.Second)); err != nil {
 				c.logger.Error().Err(err).Msg("Failed to set read deadline")
 			}
 
-			msg := c.Read()
-			switch msg := msg.(type) {
+			switch msg := c.Read().(type) {
 			case *Ping:
+				atomic.AddInt32(&count.Pings, 1)
+				c.logger.Trace().Msg("Received Ping")
+
 				if err := c.Write(&Pong{}); err != nil {
 					c.logger.Error().Err(err).Msg("Failed to write Pong response")
 				}
 			case *BlockHeaders:
-				c.logger.Info().Msgf("Received %v block headers", len(msg.BlockHeadersPacket))
+				atomic.AddInt32(&count.BlockHeaders, int32(len(msg.BlockHeadersPacket)))
+				c.logger.Trace().Msgf("Received %v BlockHeaders", len(msg.BlockHeadersPacket))
+
+				if db != nil {
+					dbCh <- struct{}{}
+					go func() {
+						db.WriteBlockHeaders(msg.BlockHeadersPacket)
+						<-dbCh
+					}()
+				}
 			case *GetBlockHeaders:
-				c.logger.Info().Interface("msg", msg).Msg("Received GetBlockHeaders request")
+				atomic.AddInt32(&count.BlockHeaderRequests, 1)
+				c.logger.Trace().Msgf("Received GetBlockHeaders request")
+
 				res := &BlockHeaders{
 					RequestId: msg.RequestId,
 				}
 				if err := c.Write(res); err != nil {
 					c.logger.Error().Err(err).Msg("Failed to write BlockHeaders response")
+					return err
 				}
 			case *BlockBodies:
-				c.logger.Info().Msgf("Received %v block bodies", len(msg.BlockBodiesPacket))
+				atomic.AddInt32(&count.BlockBodies, int32(len(msg.BlockBodiesPacket)))
```

Comment:
These counters feel like they should be openmetrics / prometheus logs.. Maybe a future improvement.